### PR TITLE
CORE-4996 consensual ledger structure

### PR DESCRIPTION
**Pull request by vlajos**
👮🏻👮🏻👮🏻 !!!! DESCRIBE YOUR CHANGES HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [ ] If you added public APIs, did you write the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the [changelog](https://docs.corda.net/head/changelog.html) (`/docs/source/changelog.rst`), and potentially the [release notes](https://docs.corda.net/head/release-notes.html) (`/docs/source/release-notes.rst`)?
- [ ] If you are contributing for the first time, please read the [contributor agreement](https://docs.corda.net/head/contributing.html) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://docs.corda.net/head/contributing.html#merging-the-changes-back-into-corda).

Thanks for your code, it's appreciated! :)

### CODE REVIEW

File: crypto/src/main/kotlin/net/corda/v5/crypto/SecureHash.kt [hunk 1]
```diff
@@ -36,6 +37,22 @@ class SecureHash(val algorithm: String, bytes: ByteArray) : OpaqueBytes(bytes) {
                 SecureHash(algorithm, data)
             }
         }
+        fun secureHash(bytes: ByteArray, algorithm: String = "SHA-256"): SecureHash {
+            val messageDigest = MessageDigest.getInstance(algorithm) // CORE-4984 review: caching messageDigest?
```
Comment (mnesbit) [inline crypto/src/main/kotlin/net/corda/v5/crypto/SecureHash.kt]: This is unfortunately not correct way to get hashers. We have a pluggable architecture for that and you need to talk to @alexey-kadyrov-r3 on correct way to get the algorithm

Comment (alexey-kadyrov-r3) [reply crypto/src/main/kotlin/net/corda/v5/crypto/SecureHash.kt]: You need to use net.corda.v5.crypto.DigestService which is OSGi component. Its methods return the SecureHash. Also the double hash is the algorithm now with the name SHA-256D

File: crypto/src/main/kotlin/net/corda/v5/crypto/merkle/MerkleTree.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package net.corda.v5.crypto.merkle
+
+import net.corda.v5.crypto.SecureHash
+
+interface MerkleTree {
+    val leaves: List<ByteArray>
+    val digestProvider: MerkleTreeHashDigestProvider
+    val depth: Int
```
Comment (mnesbit) [inline crypto/src/main/kotlin/net/corda/v5/crypto/merkle/MerkleTree.kt]: possibly don't need depth on interface, though it is used internally

**Comment by vlajos** [comment]
Continued here: https://github.com/corda/corda-api/pull/414
(depth has been removed from the interface, and digestService has been introduced.)

**Pull request closed without merging**